src/random.rs

```diff
@@ -0,0 +1,722 @@
+                stream.as_ptr(),
+            )
+        };
+        Ok(Array::from_ptr(c_array))
+    }
+}
+
+/// Sample from a categorical distribution.
+///
+/// The values are sampled from the categorical distribution specified by
+/// the unnormalized values in `logits`.   If the `shape` is not specified
+/// the result shape will be the same shape as `logits` with the `axis`
+/// dimension removed.
+///
+/// ```rust
+/// let key = mlx_rs::random::key(0);
+///
+/// let logits = mlx_rs::Array::zeros::<u32>(&[5, 20]).unwrap();
+///
+/// // produces Array of u32 shape &[5]
+/// let result = mlx_rs::random::categorical(&logits, None, None, None, &key);
+/// ```
+///
+/// - Parameters:
+///     - logits: The *unnormalized* categorical distribution(s).
+#[default_device]
+pub fn categorical_device<'a>(
+    logits: &Array,
+    axis: impl Into<Option<i32>>,
+    shape: impl Into<Option<&'a [i32]>>,
+    num_samples: impl Into<Option<i32>>,
+    key: impl Into<Option<&'a Array>>,
+    stream: StreamOrDevice,
+) -> Result<Array, Exception> {
+    let axis = axis.into().unwrap_or(-1);
+
+    let key = key.into().map_or_else(
+        || {
+            let mut state = state().lock().unwrap();
+            state.next()
+        },
+        |key| key.clone(),
+    );
+
+    let shape = shape.into();
+    let num_samples = num_samples.into();
+
+    // if we have both shape and num_samples, we need to error out
+    if shape.is_some() && num_samples.is_some() {
+        let what = "[categorical] Cannot specify both shape and num_samples";
+        let c_string = std::ffi::CString::new(what).expect("CString::new failed");
+
+        return Err(Exception { what: c_string });
+    }
+
+    if let Some(shape) = shape {
+        unsafe {
+            let c_array = try_catch_c_ptr_expr! {
+                mlx_sys::mlx_random_categorical_shape(
+                    logits.as_ptr(),
+                    axis,
+                    shape.as_ptr(),
+                    shape.len(),
+                    key.as_ptr(),
+                    stream.as_ptr(),
+                )
+            };
+            Ok(Array::from_ptr(c_array))
+        }
+    } else if let Some(num_samples) = num_samples {
+        unsafe {
+            let c_array = try_catch_c_ptr_expr! {
+                mlx_sys::mlx_random_categorical_num_samples(
+                    logits.as_ptr(),
+                    axis,
+                    num_samples,
+                    key.as_ptr(),
+                    stream.as_ptr(),
+                )
+            };
+            Ok(Array::from_ptr(c_array))
+        }
+    } else {
+        unsafe {
+            let c_array = try_catch_c_ptr_expr! {
+                mlx_sys::mlx_random_categorical(
+                    logits.as_ptr(),
+                    axis,
+                    key.as_ptr(),
+                    stream.as_ptr(),
+                )
+            };
+            Ok(Array::from_ptr(c_array))
+        }
+    }
+}
+
```

Comment:
Is `random_bits` missing? There's `mlx_random_bits` in the c binding

Reply:
I saw that but I couldn't find usage of it anywhere in the swift or python libs

Reply:
Seems like its used internally by the PRNG. I think I'd leave this out until someone needs it.